### PR TITLE
Config integration test

### PR DESCRIPTION
**Pull request by huiwangredhat**

### CODE REVIEW

File: containers/docker-ci/Dockerfile [hunk 1]
```diff
@@ -66,7 +66,6 @@ RUN patch -p0 < extract.patch && \
     patch -p0 < rpm.patch
 
 WORKDIR /opt/app-root/src
-USER 1001
```
Comment (hanchuntao) [inline containers/docker-ci/Dockerfile:69]: Why need to remove the user ?

File: requirements/devel.txt [hunk 2]
```diff
@@ -11,4 +11,8 @@ coverage==6.2
 tox-current-env==0.0.11
 # pytest dependency
 pytest-django==4.5.2
-pytest-redis==2.4.0
\ No newline at end of file
+pytest-redis==2.4.0
+# integration
+curlify
```
Comment (hanchuntao) [inline requirements/devel.txt:16]: Please provide the detail version information.

File: tests/integration/README.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+Test cases are using tox and pytest for running the tests.
+
+## Debug the integration tests
+
+Since integration tests are different with unit test and api test, it requires
+to deploy the latest patch to the openlcs-ci environment. Besides that, when
+manually debug the test cases, **make sure there are no pipeline running**.
+You should always run the commands from OpenLCS root directory.
+
+1. `oc login --server=$OCP_CONSOLE`
```
Comment (hanchuntao) [inline tests/integration/README.md:12]: `oc login --token=xxxxxx --server=https://api.mpp-w2-preprod.cfln.p1.openshiftapps.com:6443`

File: tests/integration/README.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+2. `oc project openlcs--ci`
+3. `oc rsh $gunicorn`
+5. `kinit "Your name"@IPA.REDHAT.COM`
+6. `export OPENLCS_TEST_URL="$OPENLCS_TEST_URL"`
```
Comment (hanchuntao) [inline tests/integration/README.md:16]: What is this `OPENLCS_TEST_URL` ?

File: tests/integration/conftest.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                expected_code=200, **kwargs):
+        headers = dict(headers or {})
+        if self.csrf_token:
+            headers['X-CSRFToken'] = self.csrf_token
+        if auth == 'token':
+            headers['Authorization'] = 'token {}'.format(self.get_token())
+        elif auth == 'default':
+            if OPENLCS_LOCAL_LOGIN and 'sessionid' not in self.session.cookies:
+                # Obtain CSRF token
+                self.request('/', auth=None)
```
Comment (hanchuntao) [inline tests/integration/conftest.py:35]: here auth is str type, so the default value should be ""

File: tests/integration/conftest.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            **kwargs
+        )
+        self.csrf_token = self.session.cookies.get(
+            'csrftoken',
+            default=self.csrf_token,
+        )
+        # requests internally add some headers, we don't want them in curlify
+        # output
+        response.request.headers = headers
+        print(self.to_curl(response.request))
```
Comment (hanchuntao) [inline tests/integration/conftest.py:65]: remove the print line

File: tests/integration/conftest.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        headers['Accept'] = 'application/json'
+        response = self.request(
+            endpoint=endpoint,
+            method=method,
+            headers=headers,
+            auth='token',
+            expected_code=expected_code,
+            **kwargs
+        )
+        if not return_json:
```
Comment (hanchuntao) [inline tests/integration/conftest.py:96]: `return_json` is `False`, then need a return?

File: tests/integration/test_subscription.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+            subscription_id = result['id']
+            response = client.api_call(
+                '/subscriptions/{}/'.format(subscription_id),
+                method="PATCH",
+                data={
+                    "active": "false",
+                },
+                expected_code=status.HTTP_200_OK
+            )
+    # Check the if there any active subscriptions
```
Comment (hanchuntao) [inline tests/integration/test_subscription.py:24]: Check if there exists any active subscriptions

File: tests/integration/test_subscription.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    response = client.api_call(
+        subscription_endpoint,
+        method="GET",
+        expected_code=status.HTTP_200_OK
+    )
+    # Deactivate the active subscriptions
+    for result in response['results']:
+        if result['active']:
+            subscription_id = result['id']
+            response = client.api_call(
```
Comment (hanchuntao) [inline tests/integration/test_subscription.py:16]: What is this `response` used for?

File: containers/docker-ci/Dockerfile (hunk 1 above)
Comment (huiwangredhat) [reply containers/docker-ci/Dockerfile:69]: If we have this line, it has no permission to update krb5.conf ("cp openlcs-playbooks/roles/openlcs/files/krb5.conf /etc/").

File: tests/integration/conftest.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        headers = dict(headers or {})
+        if self.csrf_token:
+            headers['X-CSRFToken'] = self.csrf_token
+        if auth == 'token':
+            headers['Authorization'] = 'token {}'.format(self.get_token())
+        elif auth == 'default':
+            if OPENLCS_LOCAL_LOGIN and 'sessionid' not in self.session.cookies:
+                # Obtain CSRF token
+                self.request('/', auth=None)
+                self.request(
```
Comment (hanchuntao) [inline tests/integration/conftest.py:36]: Why there need to request ?

File: tests/integration/conftest.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+                )
+            else:
+                kwargs['auth'] = HTTPKerberosAuth(mutual_authentication=False)
+        elif auth is None:
+            # No auth
+            pass
+        else:
+            assert False, "Unknown auth type"
+        if not urlparse(endpoint).netloc:
+            endpoint = '{}/{}'.format(OPENLCS_URL, endpoint.lstrip('/'))
```
Comment (hanchuntao) [inline tests/integration/conftest.py:50]: `endpoint = f"{OPENLCS_URL}/{endpoint.lstrip('/')}"`

File: tests/integration/README.md (hunk 3 above)
Comment (huiwangredhat) [reply tests/integration/README.md:12]: The infosec team won't allow any internal URL. I will only add the "--token=xxx", it is acceptable?

File: tests/integration/README.md (hunk 4 above)
Comment (huiwangredhat) [reply tests/integration/README.md:16]: This means https://openlcs-ci.apps.int.spoke.preprod.us-west-2.aws.paas.redhat.com/. The reason is as above.

File: containers/docker-ci/Dockerfile (hunk 1 above)
Comment (hanchuntao) [reply containers/docker-ci/Dockerfile:69]: So can we move the cp action into the Dockerfile?

File: tests/integration/README.md (hunk 3 above)
Comment (hanchuntao) [reply tests/integration/README.md:12]: Yes

File: tests/integration/conftest.py (hunk 7 above)
Comment (huiwangredhat) [reply tests/integration/conftest.py:96]: Updated.

File: tests/integration/conftest.py (hunk 6 above)
Comment (huiwangredhat) [reply tests/integration/conftest.py:65]: Removed.

File: tests/integration/conftest.py (hunk 5 above)
Comment (huiwangredhat) [reply tests/integration/conftest.py:35]: Updated.

File: requirements/devel.txt (hunk 2 above)
Comment (huiwangredhat) [reply requirements/devel.txt:16]: Updated.

File: tests/integration/test_subscription.py (hunk 8 above)
Comment (huiwangredhat) [reply tests/integration/test_subscription.py:24]: updated.

File: tests/integration/test_subscription.py (hunk 9 above)
Comment (huiwangredhat) [reply tests/integration/test_subscription.py:16]: Remove it.

File: tests/integration/conftest.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+import pytest
+import requests
+from urllib.parse import urlparse
+from requests_kerberos import HTTPKerberosAuth
+
+OPENLCS_URL = os.environ.get('OPENLCS_TEST_URL')
+OPENLCS_LOCAL_LOGIN = os.environ.get('OPENLCS_TEST_LOCAL')
+if not OPENLCS_URL:
+    OPENLCS_URL = 'http://localhost:8000'
+    OPENLCS_LOCAL_LOGIN = True
```
[the remark this answers is not in the document]
Comment (huiwangredhat) [reply tests/integration/conftest.py:11]: Updated.

File: tests/integration/conftest.py (hunk 11 above)
Comment (huiwangredhat) [reply tests/integration/conftest.py:50]: Updated.

File: tests/integration/conftest.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+                self.request(
+                    '/login/',
+                    method='POST',
+                    data=OPENLCS_LOGIN_DATA,
+                    auth=None,
+                )
+            else:
+                kwargs['auth'] = HTTPKerberosAuth(mutual_authentication=False)
+        elif auth is None:
+            # No auth
```
[the remark this answers is not in the document]
Comment (huiwangredhat) [reply tests/integration/conftest.py:45]: Updated.

File: tests/integration/conftest.py (hunk 10 above)
Comment (huiwangredhat) [reply tests/integration/conftest.py:36]: Via the request to get the CSRF token.

File: tests/integration/conftest.py (hunk 5 above)
Comment (hanchuntao) [reply tests/integration/conftest.py]: Line 40 also need change

File: tests/integration/conftest.py (hunk 10 above)
Comment (hanchuntao) [reply tests/integration/conftest.py:36]: I don't understand, why the user/password exist, we still need to get the CSRF token.

**Pull request merged**